src/Data/Text/Utf8/AhoCorasick/Automaton.hs

```diff
@@ -374,67 +379,197 @@ uAt = UVector.unsafeIndex
 -- returning a `Done`, or it can continue with a new accumulator with `Step`.
 data Next a = Done !a | Step !a
 
--- | The unit parameter here is a placeholder for a future CaseSensitivity flag,
--- see old module.
+-- | Run the automaton, possibly lowercasing the input text on the fly if case
+-- insensitivity is desired. See also `runLower`.
+--
+-- The code of this function itself is organized as a state machine as well.
+-- Each state in the diagram below corresponds to a function defined in
+-- 'runWithCase'. These functions are written in a way such that GHC identifies them
+-- as [join points](https://www.microsoft.com/en-us/research/publication/compiling-without-continuations/).
+-- This means that they can be compiled to jumps instead of function calls, which helps performance a lot.
+--
+-- @
+-- ┌────────────┐   ┌────────────────┐
+-- │consumeInput├───►followCodeUnits │
+-- └─▲──────────┘   └─▲────────────┬─┘
+--   │                │            │
+--   │              ┌─┴────────────▼─┐   ┌──────────────┐
+--   │              │lookupTransition├───►collectMatches│
+--   │              └────▲──────┬────┘   └────────────┬─┘
+--   │                   │      │                     │
+--   │                   └──────┘                     │
+--   │                                                │
+--   └────────────────────────────────────────────────┘
+-- @
+--
+-- * 'consumeInput' inspects the next code unit in the input and decides what to do with it.
+--   If necessary, it decodes a code point of up to four code units and passes them to 'followCodeUnits'.
+-- * 'followCodeUnits' pops an entry from the code unit queue and passes it to 'lookupTransition'.
+-- * 'lookupTransition' checks whether the given code unit matches any transitions at the given state.
+--   If so, it follows the transition and then loops back to 'followCodeUnits' if the code unit queue
+--   is not empty and otherwise calls 'collectMatches'.
+-- * 'collectMatches' checks whether the current state is accepting and updates the accumulator accordingly.
+--   Afterwards it loops back to 'consumeInput'.
+--
+-- NOTE: 'followCodeUnits' is actually inlined into 'consumeInput' and 'lookupTransition' by GHC.
+-- It is included in the diagram for illustrative reasons only.
+--
+-- All of these functions have the arguments 'offset', 'remaining' and 'acc' which encode the current input
```

Comment:
Haddock nit: for refering to the parameter names, does `'` work? IIRC it'll complain because it can only reference top-level identifiers that way. You could use `@` though.

(Might be best to just run `stack haddock` and see what happens)

Reply:
`'` hyperlinks an identifier and falls back to monospaced text when the identifier is not accessible, which is why I didn't notice this yet. `@thing@` will make `thing` monospace. :+1: